ClassFanOutComplexityCheck: fan-out of implicit class not measured in JEP 512 compact source files

Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/metrics/classfanoutcomplexity.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ClassFanOutComplexity">
      <property name="max" value="0"/>
      <property name="excludedClasses" value=""/>
    </module>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputClassFanOutComplexityCompact.java <<'EOF'
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;

ArrayList<String> a = new ArrayList<>();
HashMap<String, Integer> m = new HashMap<>();
LinkedList<String> l = new LinkedList<>();

void main() { }
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputClassFanOutComplexityCompact.java

Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**

ClassFanOutComplexityCheck must measure the fan-out complexity of the implicit class in a compact source file. The implicit class is a real class: it has fields, methods, and dependencies. Three references to unrelated user-visible types (`ArrayList`, `HashMap`, `LinkedList`) is a fan-out of 3 for that class and the check should fire the same violation it fires on the equivalent ordinary class.

Expected on the compact input: one violation `Class Fan-Out Complexity is 3 (max allowed is 0).`

Root cause: the check registers for `CLASS_DEF`, `INTERFACE_DEF`, `ENUM_DEF`, `RECORD_DEF`, `LITERAL_NEW`, `PACKAGE_DEF`, `IMPORT`, and uses `beginTree`/`visitToken` to track the current outer type. It only starts measuring when it enters one of the type-def tokens. A compact source file has no top-level `CLASS_DEF` for the implicit class, so the check never starts and every referenced type at the top level is silently ignored. Fix: also treat `COMPACT_COMPILATION_UNIT` as an implicit outer type for the purposes of fan-out accumulation.

**Same code inside an ordinary class:**

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > InputClassFanOutComplexityOrdinary.java <<'EOF'
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;

public class InputClassFanOutComplexityOrdinary {
    ArrayList<String> a = new ArrayList<>();
    HashMap<String, Integer> m = new HashMap<>();
    LinkedList<String> l = new LinkedList<>();

    public void main() { }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputClassFanOutComplexityOrdinary.java

Starting audit...
[ERROR] InputClassFanOutComplexityOrdinary.java:5:1: Class Fan-Out Complexity is 3 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
@@ -194,6 +194,7 @@ public abstract class AbstractClassCouplingCheck extends AbstractCheck {
                  TokenTypes.ANNOTATION_DEF,
                  TokenTypes.ENUM_DEF,
                  TokenTypes.RECORD_DEF -> visitClassDef(ast);
+            case TokenTypes.COMPACT_COMPILATION_UNIT -> visitCompactCompilationUnit(ast);
             case TokenTypes.EXTENDS_CLAUSE,
                  TokenTypes.IMPLEMENTS_CLAUSE,
                  TokenTypes.TYPE -> visitType(ast);
@@ -206,7 +207,8 @@ public abstract class AbstractClassCouplingCheck extends AbstractCheck {
 
     @Override
     public void leaveToken(DetailAST ast) {
-        if (TokenUtil.isTypeDeclaration(ast.getType())) {
+        if (TokenUtil.isTypeDeclaration(ast.getType())
+                || ast.getType() == TokenTypes.COMPACT_COMPILATION_UNIT) {
             leaveClassDef();
         }
     }
@@ -229,6 +231,16 @@ public abstract class AbstractClassCouplingCheck extends AbstractCheck {
     private void visitClassDef(DetailAST classDef) {
         final String className = classDef.findFirstToken(TokenTypes.IDENT).getText();
         createNewClassContext(className, classDef);
+    }
+
+    /**
+     * Creates new context for the implicit class declared by a compact
+     * compilation unit (JEP 512).
+     *
+     * @param compactCompilationUnit COMPACT_COMPILATION_UNIT node.
+     */
+    private void visitCompactCompilationUnit(DetailAST compactCompilationUnit) {
+        createNewClassContext("", compactCompilationUnit);
     }
 
     /** Restores previous context. */

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.java
@@ -83,6 +83,7 @@ public final class ClassFanOutComplexityCheck extends AbstractClassCouplingCheck
             TokenTypes.LITERAL_THROWS,
             TokenTypes.ANNOTATION_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
         };
     }
 

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
@@ -146,6 +146,7 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
             TokenTypes.LITERAL_THROWS,
             TokenTypes.ANNOTATION_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
         };
         assertWithMessage("Acceptable tokens should not be null")
             .that(actual)
@@ -298,6 +299,44 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
 
         verifyWithInlineConfigParser(
                 getPath("InputClassFanOutComplexitySealedClasses.java"),
+                expected);
+    }
+
+    @Test
+    public void testCompactClasses() throws Exception {
+        final String[] expected = {
+            "13:1: " + getCheckMessage(MSG_KEY, 3, 0),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("compact/InputClassFanOutComplexityCompact.java"),
+                expected);
+    }
+
+    @Test
+    public void testCompactClassesDefaultConfig() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("compact/InputClassFanOutComplexityCompactDefaultConfig.java"),
+                expected);
+    }
+
+    @Test
+    public void testCompactClassesExcludeClassesRegexps() throws Exception {
+        final String[] expected = {
+            "24:1: " + getCheckMessage(MSG_KEY, 1, 0),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("compact/InputClassFanOutComplexityCompactExcludeRegexp.java"),
+                expected);
+    }
+
+    @Test
+    public void testCompactClassesExcludedPackages() throws Exception {
+        final String[] expected = {
+            "24:1: " + getCheckMessage(MSG_KEY, 1, 0),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("compact/InputClassFanOutComplexityCompactExcluded.java"),
                 expected);
     }
 

--- a/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksCompactSourceCoverageTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksCompactSourceCoverageTest.java
@@ -105,7 +105,6 @@ public class AllChecksCompactSourceCoverageTest {
         "BooleanExpressionComplexityCheck",
         "CatchParameterNameCheck",
         "ClassDataAbstractionCouplingCheck",
-        "ClassFanOutComplexityCheck",
         "ClassMemberImpliedModifierCheck",
         "ClassTypeParameterNameCheck",
         "CommentsIndentationCheck",

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompact.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompact.java
@@ -1,0 +1,17 @@
+/*
+ClassFanOutComplexity
+max = 0
+excludedClasses = {}
+excludeClassesRegexps = (default)^$
+excludedPackages = (default)
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+// violation below 'Class Fan-Out Complexity is 3 (max allowed is 0).'
+ArrayList<String> a = new ArrayList<>();
+HashMap<String, Integer> m = new HashMap<>();
+LinkedList<String> l = new LinkedList<>();
+
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompactDefaultConfig.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompactDefaultConfig.java
@@ -1,0 +1,29 @@
+/*
+ClassFanOutComplexity
+max = (default)20
+excludedClasses = (default)ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte, \
+                  Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, \
+                  EnumSet, Exception, Float, FunctionalInterface, HashMap, HashSet, \
+                  IllegalArgumentException, IllegalStateException, IndexOutOfBoundsException, \
+                  IntStream, Integer, LinkedHashMap, LinkedHashSet, LinkedList, List, Long, \
+                  LongStream, Map, NullPointerException, Object, Optional, OptionalDouble, \
+                  OptionalInt, OptionalLong, Override, Queue, RuntimeException, SafeVarargs, \
+                  SecurityException, Set, Short, SortedMap, SortedSet, Stream, String, \
+                  StringBuffer, StringBuilder, SuppressWarnings, Throwable, TreeMap, TreeSet, \
+                  UnsupportedOperationException, Void, boolean, byte, char, double, float, \
+                  int, long, short, var, void
+excludeClassesRegexps = (default)^$
+excludedPackages = (default)
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+LocalDate today = LocalDate.now();
+LocalTime now = LocalTime.now();
+
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompactExcludeRegexp.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompactExcludeRegexp.java
@@ -1,0 +1,32 @@
+/*
+ClassFanOutComplexity
+max = 0
+excludedClasses = (default)ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte, \
+                  Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, \
+                  EnumSet, Exception, Float, FunctionalInterface, HashMap, HashSet, \
+                  IllegalArgumentException, IllegalStateException, IndexOutOfBoundsException, \
+                  IntStream, Integer, LinkedHashMap, LinkedHashSet, LinkedList, List, Long, \
+                  LongStream, Map, NullPointerException, Object, Optional, OptionalDouble, \
+                  OptionalInt, OptionalLong, Override, Queue, RuntimeException, SafeVarargs, \
+                  SecurityException, Set, Short, SortedMap, SortedSet, Stream, String, \
+                  StringBuffer, StringBuilder, SuppressWarnings, Throwable, TreeMap, TreeSet, \
+                  UnsupportedOperationException, Void, boolean, byte, char, double, float, \
+                  int, long, short, var, void
+excludeClassesRegexps = ^Local.*$
+excludedPackages = (default)
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+// violation below 'Class Fan-Out Complexity is 1 (max allowed is 0).'
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.Instant;
+
+LocalDate today = LocalDate.now();
+LocalTime now = LocalTime.now();
+Instant instant = Instant.now();
+
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompactExcluded.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/compact/InputClassFanOutComplexityCompactExcluded.java
@@ -1,0 +1,30 @@
+/*
+ClassFanOutComplexity
+max = 0
+excludedClasses = (default)ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte, \
+                  Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, \
+                  EnumSet, Exception, Float, FunctionalInterface, HashMap, HashSet, \
+                  IllegalArgumentException, IllegalStateException, IndexOutOfBoundsException, \
+                  IntStream, Integer, LinkedHashMap, LinkedHashSet, LinkedList, List, Long, \
+                  LongStream, Map, NullPointerException, Object, Optional, OptionalDouble, \
+                  OptionalInt, OptionalLong, Override, Queue, RuntimeException, SafeVarargs, \
+                  SecurityException, Set, Short, SortedMap, SortedSet, Stream, String, \
+                  StringBuffer, StringBuilder, SuppressWarnings, Throwable, TreeMap, TreeSet, \
+                  UnsupportedOperationException, Void, boolean, byte, char, double, float, \
+                  int, long, short, var, void
+excludeClassesRegexps = (default)^$
+excludedPackages = java.time
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+// violation below 'Class Fan-Out Complexity is 1 (max allowed is 0).'
+import java.time.LocalDate;
+import java.math.BigDecimal;
+
+LocalDate today = LocalDate.now();
+BigDecimal price = new BigDecimal("1.0");
+
+void main() { }

